### ISSUE TITLE
The reference assemblies for .NETFramwork.Version not found since Omnisharp upgrade

### ISSUE THREAD

**Issue by puremourning** [OmniSharp/omnisharp-roslyn#1839]
Hi there,

I maintain YouCompeteMe, a code completion server which uses omnisharp roslyn to provide c# semantic ananalysis.

We recently upgraded the bundled omnisharp from v1.34.2 to v1.35.3 and we noticed some regression test failures. We put that down to our misunderstanding, but I have since realised that it is due to a change in behaviour.

Our test project continuing the solution file and the project file is here: https://github.com/ycm-core/ycmd/tree/master/ycmd/tests/cs/testdata/testy

With v1.34.2 we get the following (successful) output:

```

[40m[32minfo[39m[22m[49m: OmniSharp.Services.DotNetCliService[0]
      DotNetPath set to dotnet
[40m[32minfo[39m[22m[49m: OmniSharp.MSBuild.Discovery.MSBuildLocator[0]
      Located 1 MSBuild instance(s)
                  1: StandAlone 16.0 - "/Users/ben/.vim/bundle/YouCompleteMe/third_party/ycmd/third_party/omnisharp-roslyn/omnisharp/.msbuild/Current/Bin"
[40m[32minfo[39m[22m[49m: OmniSharp.MSBuild.Discovery.MSBuildLocator[0]
      MSBUILD_EXE_PATH environment variable set to '/Users/ben/.vim/bundle/YouCompleteMe/third_party/ycmd/third_party/omnisharp-roslyn/omnisharp/.msbuild/Current/Bin/MSBuild.dll'
[40m[32minfo[39m[22m[49m: OmniSharp.MSBuild.Discovery.MSBuildLocator[0]
      Registered MSBuild instance: StandAlone 16.0 - "/Users/ben/.vim/bundle/YouCompleteMe/third_party/ycmd/third_party/omnisharp-roslyn/omnisharp/.msbuild/Current/Bin"
                  MSBuildExtensionsPath = /Library/Frameworks/Mono.framework/Versions/6.4.0/lib/mono/xbuild
                  BypassFrameworkInstallChecks = true
                  CscToolPath = /Users/ben/.vim/bundle/YouCompleteMe/third_party/ycmd/third_party/omnisharp-roslyn/omnisharp/.msbuild/Current/Bin/Roslyn
                  CscToolExe = csc.exe
                  MSBuildToolsPath = /Users/ben/.vim/bundle/YouCompleteMe/third_party/ycmd/third_party/omnisharp-roslyn/omnisharp/.msbuild/Current/Bin
                  TargetFrameworkRooTargetFrameworkRootPathtPath = /Library/Frameworks/Mono.framework/Versions/6.4.0/lib/mono/xbuild-frameworks
[40m[32minfo[39m[22m[49m: OmniSharp.Http.Startup[0]
      Starting OmniSharp on MacOS 10.15.5 (x64)
[40m[32minfo[39m[22m[49m: OmniSharp.Cake.CakeProjectSystem[0]
      Detecting Cake files in '/Users/ben/.vim/bundle/YouCompleteMe/third_party/ycmd/ycmd/tests/cs/testdata/testy'.
[40m[32minfo[39m[22m[49m: OmniSharp.Cake.CakeProjectSystem[0]
      Could not find any Cake files
[40m[32minfo[39m[22m[49m: OmniSharp.MSBuild.ProjectSystem[0]
      Detecting projects in '/Users/ben/.vim/bundle/YouCompleteMe/third_party/ycmd/ycmd/tests/cs/testdata/testy/testy.sln'.
[40m[32minfo[39m[22m[49m: OmniSharp.MSBuild.ProjectManager[0]
      Queue project update for '/Users/ben/.vim/bundle/YouCompleteMe/third_party/ycmd/ycmd/tests/cs/testdata/testy/testy.csproj'
[40m[32minfo[39m[22m[49m: OmniSharp.Script.ScriptProjectSystem[0]
      Detecting CSX files in '/Users/ben/.vim/bundle/YouCompleteMe/third_party/ycmd/ycmd/tests/cs/testdata/testy'.
[40m[32minfo[39m[22m[49m: OmniSharp.Script.ScriptProjectSystem[0]
      Could not find any CSX files
[40m[32minfo[39m[22m[49m: OmniSharp.WorkspaceInitializer[0]
      Invoking Workspace Options Provider: OmniSharp.Roslyn.CSharp.Services.CSharpFormattingWorkspaceOptionsProvider, Order: 0
[40m[32minfo[39m[22m[49m: OmniSharp.MSBuild.ProjectManager[0]
      Loading project: /Users/ben/.vim/bundle/YouCompleteMe/third_party/ycmd/ycmd/tests/cs/testdata/testy/testy.csproj
[40m[32minfo[39m[22m[49m: OmniSharp.WorkspaceInitializer[0]
      Invoking Workspace Options Provider: OmniSharp.Roslyn.CSharp.Services.RenameWorkspaceOptionsProvider, Order: 100
[40m[32minfo[39m[22m[49m: OmniSharp.WorkspaceInitializer[0]
      Invoking Workspace Options Provider: OmniSharp.Roslyn.CSharp.Services.EditorConfigWorkspaceOptionsProvider, Order: 200
[40m[32minfo[39m[22m[49m: OmniSharp.WorkspaceInitializer[0]
      Configuration finished.
[40m[32minfo[39m[22m[49m: OmniSharp.Http.Startup[0]
      Omnisharp server running on port '56398' at location '/Users/ben/.vim/bundle/YouCompleteMe/third_party/ycmd/ycmd/tests/cs/testdata/testy' on host -1.
[40m[32minfo[39m[22m[49m: OmniSharp.MSBuild.ProjectManager[0]
      Successfully loaded project file '/Users/ben/.vim/bundle/YouCompleteMe/third_party/ycmd/ycmd/tests/cs/testdata/testy/testy.csproj'.
```

However with the only change being to upgrade to v1.35.3, we get the following error, and semantic analysis is constrained to the current file:

```
[40m[32minfo[39m[22m[49m: OmniSharp.Services.DotNetCliService[0]
      DotNetPath set to dotnet
[40m[32minfo[39m[22m[49m: OmniSharp.MSBuild.Discovery.MSBuildLocator[0]
      Located 1 MSBuild instance(s)
                  1: StandAlone 16.4 - "/Users/ben/.vim/bundle/YouCompleteMe/third_party/ycmd/third_party/omnisharp-roslyn/omnisharp/.msbuild/Current/Bin"
[40m[32minfo[39m[22m[49m: OmniSharp.MSBuild.Discovery.MSBuildLocator[0]
      MSBUILD_EXE_PATH environment variable set to '/Users/ben/.vim/bundle/YouCompleteMe/third_party/ycmd/third_party/omnisharp-roslyn/omnisharp/.msbuild/Current/Bin/MSBuild.dll'
[40m[32minfo[39m[22m[49m: OmniSharp.MSBuild.Discovery.MSBuildLocator[0]
      Registered MSBuild instance: StandAlone 16.4 - "/Users/ben/.vim/bundle/YouCompleteMe/third_party/ycmd/third_party/omnisharp-roslyn/omnisharp/.msbuild/Current/Bin"
                  CscToolExe = csc.exe
                  MSBuildToolsPath = /Users/ben/.vim/bundle/YouCompleteMe/third_party/ycmd/third_party/omnisharp-roslyn/omnisharp/.msbuild/Current/Bin
                  CscToolPath = /Users/ben/.vim/bundle/YouCompleteMe/third_party/ycmd/third_party/omnisharp-roslyn/omnisharp/.msbuild/Current/Bin/Roslyn
                  BypassFrameworkInstallChecks = true
                  MSBuildExtensionsPath = /Users/ben/.vim/bundle/YouCompleteMe/third_party/ycmd/third_party/omnisharp-roslyn/omnisharp/.msbuild
[40m[32minfo[39m[22m[49m: OmniSharp.Http.Startup[0]
      Starting OmniSharp on MacOS 10.15.5 (x64)
[40m[32minfo[39m[22m[49m: OmniSharp.Cake.CakeProjectSystem[0]
      Detecting Cake files in '/Users/ben/.vim/bundle/YouCompleteMe/third_party/ycmd/ycmd/tests/cs/testdata/testy'.
[40m[32minfo[39m[22m[49m: OmniSharp.Cake.CakeProjectSystem[0]
      Could not find any Cake files
[40m[32minfo[39m[22m[49m: OmniSharp.MSBuild.ProjectSystem[0]
      Detecting projects in '/Users/ben/.vim/bundle/YouCompleteMe/third_party/ycmd/ycmd/tests/cs/testdata/testy/testy.sln'.
[40m[32minfo[39m[22m[49m: OmniSharp.MSBuild.ProjectManager[0]
      Queue project update for '/Users/ben/.vim/bundle/YouCompleteMe/third_party/ycmd/ycmd/tests/cs/testdata/testy/testy.csproj'
[40m[32minfo[39m[22m[49m: OmniSharp.Script.ScriptProjectSystem[0]
      Detecting CSX files in '/Users/ben/.vim/bundle/YouCompleteMe/third_party/ycmd/ycmd/tests/cs/testdata/testy'.
[40m[32minfo[39m[22m[49m: OmniSharp.Script.ScriptProjectSystem[0]
      Could not find any CSX files
[40m[32minfo[39m[22m[49m: OmniSharp.WorkspaceInitializer[0]
      Invoking Workspace Options Provider: OmniSharp.Roslyn.CSharp.Services.CSharpFormattingWorkspaceOptionsProvider, Order: 0
[40m[32minfo[39m[22m[49m: OmniSharp.MSBuild.ProjectManager[0]
      Loading project: /Users/ben/.vim/bundle/YouCompleteMe/third_party/ycmd/ycmd/tests/cs/testdata/testy/testy.csproj
[40m[32minfo[39m[22m[49m: OmniSharp.WorkspaceInitializer[0]
      Invoking Workspace Options Provider: OmniSharp.Roslyn.CSharp.Services.RenameWorkspaceOptionsProvider, Order: 100
[40m[32minfo[39m[22m[49m: OmniSharp.WorkspaceInitializer[0]
      Invoking Workspace Options Provider: OmniSharp.Roslyn.CSharp.Services.ImplementTypeWorkspaceOptionsProvider, Order: 110
[40m[32minfo[39m[22m[49m: OmniSharp.WorkspaceInitializer[0]
      Invoking Workspace Options Provider: OmniSharp.Roslyn.CSharp.Services.BlockStructureWorkspaceOptionsProvider, Order: 140
[40m[32minfo[39m[22m[49m: OmniSharp.WorkspaceInitializer[0]
      Configuration finished.
[40m[32minfo[39m[22m[49m: OmniSharp.Http.Startup[0]
      Omnisharp server running on port '56483' at location '/Users/ben/.vim/bundle/YouCompleteMe/third_party/ycmd/ycmd/tests/cs/testdata/testy' on host -1.
[41m[30mfail[39m[22m[49m: OmniSharp.MSBuild.ProjectLoader[0]
      The reference assemblies for .NETFramework,Version=v4.0 were not found. To resolve this, install the Developer Pack (SDK/Targeting Pack) for this framework version or retarget your application. You can download .NET Framework Developer Packs at https://aka.ms/msbuild/developerpacks
```

I saw on a few other issues that there's an environment variable that can be set to override the framework location, but I can't ask my users to set that it seems like a regression.

Is there something we're missing? Is there a resolution that doesn't involve a change in all users' environments?

OS: MacOS, Linux

**Comment by puremourning** [comment]
Correction - perhaps unrelated but the same test failed on Windows too but with a different error:

```
info: OmniSharp.Services.DotNetCliService[0]
      DotNetPath set to dotnet
info: OmniSharp.MSBuild.Discovery.MSBuildLocator[0]
      Located 2 MSBuild instance(s)
                  1: Visual Studio Community 2019 16.4.29613.14 - "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Current\Bin"
                  2: StandAlone 16.0 - "C:\Users\ben\.vim\bundle\YouCompleteMe-Clean\third_party\ycmd\third_party\omnisharp-roslyn\.msbuild\Current\Bin"
info: OmniSharp.MSBuild.Discovery.MSBuildLocator[0]
      Registered MSBuild instance: Visual Studio Community 2019 16.4.29613.14 - "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Current\Bin"
info: OmniSharp.Http.Startup[0]
      Starting OmniSharp on Windows 6.2.9200.0 (x64)
info: OmniSharp.Cake.CakeProjectSystem[0]
      Detecting Cake files in 'H:\vimfiles\bundle\YouCompleteMe\third_party\ycmd\ycmd\tests\cs\testdata\testy'.
info: OmniSharp.Cake.CakeProjectSystem[0]
      Could not find any Cake files
info: OmniSharp.MSBuild.ProjectSystem[0]
      Detecting projects in 'H:\vimfiles\bundle\YouCompleteMe\third_party\ycmd\ycmd\tests\cs\testdata\testy\testy.sln'.
info: OmniSharp.MSBuild.ProjectManager[0]
      Queue project update for 'H:\vimfiles\bundle\YouCompleteMe\third_party\ycmd\ycmd\tests\cs\testdata\testy\testy.csproj'
info: OmniSharp.Script.ScriptProjectSystem[0]
      Detecting CSX files in 'H:\vimfiles\bundle\YouCompleteMe\third_party\ycmd\ycmd\tests\cs\testdata\testy'.
info: OmniSharp.Script.ScriptProjectSystem[0]
      Could not find any CSX files
info: OmniSharp.WorkspaceInitializer[0]
      Invoking Workspace Options Provider: OmniSharp.Roslyn.CSharp.Services.CSharpFormattingWorkspaceOptionsProvider, Order: 0
info: OmniSharp.MSBuild.ProjectManager[0]
      Loading project: H:\vimfiles\bundle\YouCompleteMe\third_party\ycmd\ycmd\tests\cs\testdata\testy\testy.csproj
info: OmniSharp.WorkspaceInitializer[0]
      Invoking Workspace Options Provider: OmniSharp.Roslyn.CSharp.Services.RenameWorkspaceOptionsProvider, Order: 100
info: OmniSharp.WorkspaceInitializer[0]
      Invoking Workspace Options Provider: OmniSharp.Roslyn.CSharp.Services.EditorConfigWorkspaceOptionsProvider, Order: 200
info: OmniSharp.WorkspaceInitializer[0]
      Configuration finished.
info: OmniSharp.Http.Startup[0]
      Omnisharp server running on port '53072' at location 'H:\vimfiles\bundle\YouCompleteMe\third_party\ycmd\ycmd\tests\cs\testdata\testy' on host -1.
info: OmniSharp.OmniSharpWorkspace[0]
      Miscellaneous file: H:\vimfiles\bundle\YouCompleteMe\third_party\ycmd\ycmd\tests\cs\testdata\testy\GotoTestCase.cs added to workspace
fail: Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware[1]
      An unhandled exception has occurred while executing the request.
System.NullReferenceException: Object reference not set to an instance of an object.
   at OmniSharp.Roslyn.CSharp.Services.Navigation.FindImplementationsService.<Handle>d__2.MoveNext() in C:\projects\omnisharp-roslyn\src\OmniSharp.Roslyn.CSharp\Services\Navigation\FindImplementationsService.cs:line 69
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at OmniSharp.Endpoint.EndpointHandler`2.<HandleSingleRequest>d__18.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at OmniSharp.Endpoint.EndpointHandler`2.<Process>d__16.MoveNext() in C:\projects\omnisharp-roslyn\src\OmniSharp.Host\Endpoint\EndpointHandler.cs:line 127
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at OmniSharp.Http.Middleware.EndpointMiddleware.<Invoke>d__7.MoveNext() in C:\projects\omnisharp-roslyn\src\OmniSharp.Http\Middleware\EndpointMiddleware.cs:line 91
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware.<Invoke>d__6.MoveNext()
info: OmniSharp.MSBuild.ProjectManager[0]
      Successfully loaded project file 'H:\vimfiles\bundle\YouCompleteMe\third_party\ycmd\ycmd\tests\cs\testdata\testy\testy.csproj'.
info: OmniSharp.Http.Middleware.LoggingMiddleware[0]
      /findimplementations: 500 2252ms
info: OmniSharp.MSBuild.ProjectManager[0]
      Adding project 'H:\vimfiles\bundle\YouCompleteMe\third_party\ycmd\ycmd\tests\cs\testdata\testy\testy.csproj'
info: OmniSharp.MSBuild.ProjectManager[0]
      Update project: testy
fail: Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware[1]
      An unhandled exception has occurred while executing the request.
System.ArgumentException: '<Document52c76326-c3cf-4a52-a796-4c72275b5beb>' is not part of the workspace.
   at Microsoft.CodeAnalysis.Workspace.CheckDocumentIsInCurrentSolution(DocumentId documentId)
   at Microsoft.CodeAnalysis.Workspace.OnAnyDocumentTextChanged(DocumentId documentId, SourceText newText, PreservationMode mode, Action`1 checkIsInCurrentSolution, Func`3 getRelatedDocuments, Func`5 updateSolutionWithText, WorkspaceChangeKind changeKind, Boolean isCodeDocument)
   at Microsoft.CodeAnalysis.Workspace.OnDocumentTextChanged(DocumentId documentId, SourceText newText, PreservationMode mode)
   at Microsoft.CodeAnalysis.Workspace.ApplyChangedDocument(ProjectChanges projectChanges, DocumentId documentId)
   at Microsoft.CodeAnalysis.Workspace.ApplyProjectChanges(ProjectChanges projectChanges)
   at OmniSharp.OmniSharpWorkspace.ApplyProjectChanges(ProjectChanges projectChanges) in C:\projects\omnisharp-roslyn\src\OmniSharp.Roslyn\OmniSharpWorkspace.cs:line 492
   at Microsoft.CodeAnalysis.Workspace.TryApplyChanges(Solution newSolution, IProgressTracker progressTracker)
   at OmniSharp.Roslyn.BufferManager.<UpdateBufferAsync>d__8.MoveNext() in C:\projects\omnisharp-roslyn\src\OmniSharp.Roslyn\BufferManager.cs:line 90
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at OmniSharp.Roslyn.CSharp.Services.Buffer.UpdateBufferService.<Handle>d__2.MoveNext() in C:\projects\omnisharp-roslyn\src\OmniSharp.Roslyn.CSharp\Services\Buffer\UpdateBufferService.cs:line 23
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at OmniSharp.Endpoint.EndpointHandler`2.<HandleSingleRequest>d__18.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at OmniSharp.Endpoint.EndpointHandler`2.<Process>d__16.MoveNext() in C:\projects\omnisharp-roslyn\src\OmniSharp.Host\Endpoint\EndpointHandler.cs:line 127
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at OmniSharp.Endpoint.EndpointHandler`2.<Process>d__16.MoveNext() in C:\projects\omnisharp-roslyn\src\OmniSharp.Host\Endpoint\EndpointHandler.cs:line 110
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at OmniSharp.Http.Middleware.EndpointMiddleware.<Invoke>d__7.MoveNext() in C:\projects\omnisharp-roslyn\src\OmniSharp.Http\Middleware\EndpointMiddleware.cs:line 91
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware.<Invoke>d__6.MoveNext()
info: OmniSharp.Http.Middleware.LoggingMiddleware[0]
      /gotodefinition: 500 136ms
```

**Comment by puremourning** [comment]
FWIW the same test does detect VS correctly on my windows VM:

```
info: OmniSharp.Services.DotNetCliService[0]
      DotNetPath set to dotnet
info: OmniSharp.MSBuild.Discovery.MSBuildLocator[0]
      Located 2 MSBuild instance(s)
                  1: Visual Studio Community 2019 16.4.29613.14 - "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Current\Bin"
                  2: StandAlone 16.4 - "C:\Users\ben\.vim\bundle\YouCompleteMe\third_party\ycmd\third_party\omnisharp-roslyn\.msbuild\Current\Bin"
info: OmniSharp.MSBuild.Discovery.MSBuildLocator[0]
      Registered MSBuild instance: Visual Studio Community 2019 16.4.29613.14 - "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Current\Bin"
info: OmniSharp.Http.Startup[0]
      Starting OmniSharp on Windows 6.2.9200.0 (x64)
info: OmniSharp.Cake.CakeProjectSystem[0]
      Detecting Cake files in 'H:\vimfiles\bundle\YouCompleteMe\third_party\ycmd\ycmd\tests\cs\testdata\testy'.
info: OmniSharp.Cake.CakeProjectSystem[0]
      Could not find any Cake files
info: OmniSharp.MSBuild.ProjectSystem[0]
      Detecting projects in 'H:\vimfiles\bundle\YouCompleteMe\third_party\ycmd\ycmd\tests\cs\testdata\testy\testy.sln'.
info: OmniSharp.MSBuild.ProjectManager[0]
      Queue project update for 'H:\vimfiles\bundle\YouCompleteMe\third_party\ycmd\ycmd\tests\cs\testdata\testy\testy.csproj'
info: OmniSharp.Script.ScriptProjectSystem[0]
      Detecting CSX files in 'H:\vimfiles\bundle\YouCompleteMe\third_party\ycmd\ycmd\tests\cs\testdata\testy'.
info: OmniSharp.Script.ScriptProjectSystem[0]
      Could not find any CSX files
info: OmniSharp.WorkspaceInitializer[0]
      Invoking Workspace Options Provider: OmniSharp.Roslyn.CSharp.Services.CSharpFormattingWorkspaceOptionsProvider, Order: 0
info: OmniSharp.MSBuild.ProjectManager[0]
      Loading project: H:\vimfiles\bundle\YouCompleteMe\third_party\ycmd\ycmd\tests\cs\testdata\testy\testy.csproj
info: OmniSharp.WorkspaceInitializer[0]
      Invoking Workspace Options Provider: OmniSharp.Roslyn.CSharp.Services.RenameWorkspaceOptionsProvider, Order: 100
info: OmniSharp.WorkspaceInitializer[0]
      Invoking Workspace Options Provider: OmniSharp.Roslyn.CSharp.Services.ImplementTypeWorkspaceOptionsProvider, Order: 110
info: OmniSharp.WorkspaceInitializer[0]
      Invoking Workspace Options Provider: OmniSharp.Roslyn.CSharp.Services.BlockStructureWorkspaceOptionsProvider, Order: 140
info: OmniSharp.WorkspaceInitializer[0]
      Configuration finished.
info: OmniSharp.Http.Startup[0]
      Omnisharp server running on port '53153' at location 'H:\vimfiles\bundle\YouCompleteMe\third_party\ycmd\ycmd\tests\cs\testdata\testy' on host -1.
info: OmniSharp.MSBuild.ProjectManager[0]
      Successfully loaded project file 'H:\vimfiles\bundle\YouCompleteMe\third_party\ycmd\ycmd\tests\cs\testdata\testy\testy.csproj'.
info: OmniSharp.MSBuild.ProjectManager[0]
      Adding project 'H:\vimfiles\bundle\YouCompleteMe\third_party\ycmd\ycmd\tests\cs\testdata\testy\testy.csproj'
info: OmniSharp.MSBuild.ProjectManager[0]
      Update project: testy
```

**Comment by filipw** [comment]
What you experience on macOS/Linux is by design. In the "old" log you are using stand alone msbuild of OmniSharp, and it mistakenly used bits from your global Mono:

```
MSBuildExtensionsPath = /Library/Frameworks/Mono.framework/Versions/6.4.0/lib/mono/xbuild
```

The new log shows that we don't do it anymore (and this is by design):

```
MSBuildExtensionsPath = /Users/ben/.vim/bundle/YouCompleteMe/third_party/ycmd/third_party/omnisharp-roslyn/omnisharp/.msbuild
```

See https://github.com/OmniSharp/omnisharp-roslyn/pull/1656 for more details.

In order to use reference assemblies for desktop .NET on macOS/Linux you need a proper Mono installation (for example from https://www.mono-project.com/download/stable/). You can also see how we install it in our build script https://github.com/OmniSharp/omnisharp-roslyn/blob/master/azure-pipelines.yml#L58-L66 I don't know how you bootstrap OmniSharp server, but OmniSharp should then discover global Mono using this logic https://github.com/OmniSharp/omnisharp-roslyn/blob/master/src/OmniSharp.Host/MSBuild/Discovery/MSBuildLocator.cs#L44

**Comment by puremourning** [comment]
Forgive my ignorance, but what's mistaken about using the mono in /Library/Framworks .. is that not a proper mono ? I mean it used to work fine.

I also have mono installed from homebrew but was unable to get omnisharp t use that either, even if it were first in the path.

My colleague @bstaletic sees the same behaviour on linux after doing `pacman -s mono`. 

> don't know how you bootstrap OmniSharp server

We just [extract it to a directory](https://github.com/ycm-core/ycmd/blob/663025ba9f761282f2c3af4e0424cc3c5f2302fd/build.py#L774) and [use the `run` script](https://github.com/ycm-core/ycmd/blob/663025ba9f761282f2c3af4e0424cc3c5f2302fd/ycmd/completers/cs/cs_completer.py#L41-L50)

Apologies for my vagueness, the contributor who maintains our omnisharp integration is not available, so I'm having to sort of learn this stuff as i go.

Anyway, in summary is the position that omnisharp requires the "official" mono distribution (https://www.mono-project.com/download/stable/) only and we should inform our users to install that if they have problems (and do that in our CI) ?

**Comment by puremourning** [comment]
just found [this comment](https://github.com/OmniSharp/omnisharp-roslyn/blob/master/src/OmniSharp.Host/MSBuild/Discovery/Providers/MonoInstanceProvider.cs#L24-L31) which maybe explains it.

I think i have an idea of how to proceed:

- If using the embedded mono in omnisharp (the run script), only the official packages are supported
- otherwise, you can run omnisharp under any other mono and it will pull the reference assemblies from there

Does that sound about right ?

For example if i were to run my mono from /Library/Frameworks/Mono and run "OmniSharp.exe" , you'd expect it to find the reference assemblies ?

**Comment by filipw** [comment]
yes, sorry if my comment was confusing.

 -  when starting OmniSharp on the embedded Mono, no interaction with global Mono happens. It used to happen, so your scenario worked "by accident" but it was really a bug that we fixed. On embedded Mono, you will encounter missing Desktop .NET reference assemblies
 - when starting OmniSharp on the global/external Mono, there is a discovery process for MsBuild, which would discover global Mono and its reference assemblies

This is by design - since the 1st approach is entirely self contained now like it was intended all along.
So you want to switch to the 2nd approach to have your scenario working again.

Hope this helps.

**Comment by filipw** [comment]
I will close this issue as I assume the explanation makes sense. Please reopen if things still don't work.
